blueprints/public.py

```diff
@@ -1,42 +1,119 @@
-from flask import Blueprint, render_template, request, send_file, make_response
+from flask import Blueprint, render_template, request
 from modules.db import db
 from modules.email import send_confirmation
 from bson.objectid import ObjectId
 from icalendar import Calendar, Event
 from datetime import datetime
+import uuid
+
 
 public = Blueprint('public', __name__, template_folder='../templates', static_folder='../static/build/')
 
-@public.route('/')
+@public.route('/', methods=['GET'])
 def public_index():
-    return render_template('home.html')
+    if request.method == "GET":
+        return render_template('home.html')
 
 @public.route('/add', methods=['POST', 'GET'])
 def add_event():
     if request.method == 'GET':
         return render_template('add.html')
+    elif request.method == 'POST':
+        magicid = uuid.uuid4()
+        eventid = ObjectId
+        event = {
+            "_id": eventid,
+            "title": request.form["title"],
+            "description": request.form["description"],
+            "location": request.form["location"],
+            "start": request.form["start"],
+            "end": request.form["end"],
+            "rrule": request.form["rrule"],
+            "last_edited": datetime.now(timezon.utc),
+            "magic": magicid,
+        }
+        db.events.insert_one(event)
+
+        link = request.base_url + "/edit?event_id=" + eventid + "&magic=" + magicid
+        send_confirmation(request.form['email'], link)
+        return redirect(url_for('confirmation/' + eventid))
 
 @public.route('/about', methods=['GET'])
 def about_page():
-    return render_template("about.html")
+    if request.method == "GET":
+        return render_template("about.html")
 
-@public.route('/<page>')
+@public.route('/<page>', methods=['GET'])
 def public_page(page):
-    try:
-        rendered_page = render_template("%s.html" % page)
-    except:
-        rendered_page = render_template("404.html")
+    if request.method == "GET":
+        try:
+            rendered_page = render_template("%s.html" % page)
+        except:
+            rendered_page = render_template("404.html")
+        return rendered_page
 
-    return rendered_page
+@public.route('/edit/<event_id>', methods=['PUT', 'GET', 'DELETE'])
+def edit_event(event_id):
+    
+    magicid = request.args.get('magic')
+    eventid = event_id
+    event_data = db.events.find_one({'_id' : ObjectId(eventid)})
+    if not event_data['magic'] == magicid or event_data is None:
+        return render_template("404.html")
+        #render a customized error page eventually?
```

Comment:
Eventually I will make this a 403 error (not authorized), but no need to change for now.